### PR TITLE
Add user backend and database setup, add user fields to room.

### PR DESCRIPTION
**Pull request by zpChris**
### List of Changes:
- Enable ability to add user objects to database, creating the relevant files: User DAO, DTO, Repository, Controller, Model, and Service.
- Add a User "Host" and Set of User Objects to the Room class.
  - When creating a room, you now must include a User "Host".
  - There are `OneToOne(...)` and `ManyToOne(...)` annotations on the User fields; here is [a tutorial on the subject](https://vladmihalcea.com/the-best-way-to-map-a-onetomany-association-with-jpa-and-hibernate/). This caused a lot of strife for me when developing this -- I'd recommend researching this at least a bit to understand the different parameters for these annotations.
  - Add new `ERROR_USER_ALREADY_PRESENT` and `ERROR_NO_USER_FOUND` errors to the `JoinRoomResponse`.
- Revise Room tests and add new User tests (both `UserService` and `MockMvc`).

### Notes:
- **Important note:** This branch is currently set to merge into `SocketUserConnections`, not `master`. Thus, it does not run any checks. We will want to make sure these checks are run before merging (so review my PR #26 first).
- This PR has a lot of changes (**a lot of which are tests**) , and I think it may be useful to do a run-through of the changes. Sorry about this -- I had trouble splitting this up further, as most of the changes mirror @alankbi's prior PR, #17. If you understand his, you already understand the basis for much of mine.
- I added `protected RoomMapper() {}` to the `RoomMapper` class to address the **sonarlint** warning that "Utility classes should not have public constructors", as referenced in [this prior discussion](https://github.com/rocketden/main/pull/23/files/a8e9ac71600e1200600dd45a4ce9553e8374f1ea#r486367158). I also added this to the new `UserMapper` class.
- Eventually, we will want to remove users from rooms, and delete users and rooms from the database; however, we can add those functions later (as the PR is already fairly bloated).
- If you are checking current table contents through **Sequel Pro**, in order to delete the elements in the room, room-users (this holds the `Set<User>` field on the `Room` class), and user tables, as far as I'm aware, you need to force delete all the tables themselves. After running the code again, these tables will reappear.

### Screencast and Images:

[Screencast of the REST Calls in Postman](https://www.loom.com/share/b161df3ff97d4260bbadf16e2aa3f7a8)

There are no changes to the UI on this PR, so there are no images attached.

### CODE REVIEW

File: src/main/java/com/rocketden/main/dto/user/UserMapper.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/dto/user/UserMapper.java:13]: Suggestion: add tests for this class

File: src/main/java/com/rocketden/main/dto/room/RoomMapper.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/dto/room/RoomMapper.java:12]: Suggestion: add tests for this class (verify additional fields are being mapped correctly)

File: src/main/java/com/rocketden/main/model/Room.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/model/Room.java:37]: I saw the [link](https://vladmihalcea.com/the-best-way-to-map-a-onetomany-association-with-jpa-and-hibernate/) you shared and it suggested not to have a unidirectional OneToMany mapping. Maybe we could change this to the bidirectional mapping (i.e. remove JoinColumn, add mappedBy to the OneToMany annotation, and add field in User for room) suggested in that article? However, this requires the extra step of adding custom methods to ensure synchronization on both ends, as well as overriding equals and hashCode in some special format.

Another option would be similar to that article's last solution with just a ManyToOne relationship in the User class, and then adding some kind of findUsersByRoom method to UserRepository. 

This is a complex topic so we might need to talk about this in more detail.

File: src/main/java/com/rocketden/main/model/Room.java
Comment (alankbi) [inline src/main/java/com/rocketden/main/model/Room.java:33]: Maybe this can be a OneToOne relationship? That might simplify the use case, and also it might make sense that a user can't host two rooms at the same time.

**Review by alankbi — CHANGES_REQUESTED** [review]
Great output on getting all this done; aside from a couple nits, most of my comments are literally just asking questions about future design haha. Some of these might be worth discussing before merging the PR tho so I requested changes for now (but a lot of those might end up being resolved after talking about it).

**Comment by zpChris** [comment]
[Add and sync user to backend](https://trello.com/c/eVOWU2UJ/103-add-and-sync-user-to-backend)

File: src/main/java/com/rocketden/main/dto/room/CreateRoomResponse.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/dto/room/CreateRoomResponse.java:16]: No, I don't think we need to return the host with this response; and yes, I think your intuition is right that we should be returning the `RoomDto` object -- that makes more sense, as it gives us all the relevant information that we may want. Do you think I should update that in this PR?

File: src/main/java/com/rocketden/main/service/RoomService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/RoomService.java:76]: You're right, it did -- done!

File: src/main/java/com/rocketden/main/service/RoomService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/RoomService.java:106]: Absolutely, that completely makes sense. Thanks for noting this -- I think commenting these things will really help us for future design, much appreciated.

File: src/main/java/com/rocketden/main/service/RoomService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/RoomService.java:111]: Done!

File: src/main/java/com/rocketden/main/service/UserService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/UserService.java:23]: > Is this route/method currently used? If not, are we planning on allow people to create actual accounts or require hosts to do so for the MVP?
It is not used beyond testing -- I included this initially because I thought I would first need to create the User object, _then_ add them to the database. However, I was wrong, so I'm happy to remove this -- especially because it would cut down on the relevant code. I don't think we are requiring any actual accounts for the MVP.

> it seems like we might end up with both Users and AnonymousUsers and might need to figure out how to distinguish them
This is a fantastic idea, 100% agree.

File: src/main/java/com/rocketden/main/model/Room.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/model/Room.java:37]: I think it makes sense to do this -- especially because seeing what `Room` a User is currently a member of would be a useful field. I think I'd prefer to update this in a second PR though; I don't want to bloat this PR too much more, and there are a good number of changes required to fully implement this. I added it as a [Trello task here](https://trello.com/c/jWVjcKix/119-change-users-field-in-the-room-object-to-a-bidirectional-mapping), let me know what you think.

File: src/main/java/com/rocketden/main/model/Room.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/model/Room.java]: This makes sense, thanks for noting that -- done!

File: src/main/java/com/rocketden/main/dto/room/RoomMapper.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/dto/room/RoomMapper.java:12]: Done!

File: src/main/java/com/rocketden/main/dto/user/UserMapper.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/dto/user/UserMapper.java:13]: Done!

File: src/main/java/com/rocketden/main/model/User.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/model/User.java:46]: This is a great suggestion; done, thank you!

File: src/main/java/com/rocketden/main/model/Room.java
Comment (alankbi) [reply src/main/java/com/rocketden/main/model/Room.java:37]: Sounds good 👍

File: src/main/java/com/rocketden/main/dto/room/CreateRoomResponse.java
Comment (alankbi) [reply src/main/java/com/rocketden/main/dto/room/CreateRoomResponse.java:16]: Just wanted to leave a note about that and I added a card for it in the backlog, so I think for this PR it's all good

File: src/main/java/com/rocketden/main/service/UserService.java
Comment (alankbi) [reply src/main/java/com/rocketden/main/service/UserService.java:23]: My personal leaning might be just to leave it as is in case we delete it but end up wanting to pursue it in the near future. However, it can also help to cut down on lines of code, so I'll leave this up to you

**Review by alankbi — APPROVED** [review]
Looks good! There's just one open conversation left which I'll let you decide on.

File: src/main/java/com/rocketden/main/service/UserService.java
Comment (zpChris) [reply src/main/java/com/rocketden/main/service/UserService.java:23]: Ok, I think that's fair, I'll leave this in 👍

**Review by helenli8181 — APPROVED** [review]
Thank you for making the demo video! That was super helpful and I was able to replicate the results locally. All the code looks good to me :)

**Pull request merged**
